webpackJsonp([6],{"/TYz":function(e,a,t){"use strict";function s(e){t("a3dU")}Object.defineProperty(a,"__esModule",{value:!0});var n=t("sFlg"),r=t("K2ND"),i=t("VU/8"),o=s,l=i(n.a,r.a,!1,o,null,null);a.default=l.exports},K2ND:function(e,a,t){"use strict";var s=function(){var e=this,a=e.$createElement;e._self._c;return e._m(0)},n=[function(){var e=this,a=e.$createElement,t=e._self._c||a;return t("div",{staticClass:"sans-serif"},[t("article",{staticClass:"cf ph3 ph5-ns pv5"},[t("header",{staticClass:"fn fl-ns w-50-ns pr4-ns"},[t("h1",{staticClass:"mb3 mt0 lh-title"},[e._v("\n        Integrated \n        Differential \n        Expression and \n        Pathway analysis")]),t("time",{staticClass:"f6 ttu tracked gray"},[e._v("Ge-Lab")])]),t("div",{staticClass:"fn fl-ns w-50-ns"},[t("p",{staticClass:"lh-copy w-100"},[e._v("\n        Integrated Differential Expression and Pathway analysis (iDEP) of transcriptomic data. See documentation and manuscript. \n        Based on annotation of 69 metazoa and 42 plant genomes in Ensembl BioMart as of 6/4/2017. \n        Additional data from KEGG, Reactome, MSigDB (human), GSKB (mouse) and araPath (arabidopsis). \n        For feedbacks or data contributions (genes and GO mapping of any species), \n        please contact us, or visit our homepage. Send us suggestions or any error message to help improve iDEP.\n      ")]),t("p",{staticClass:"lh-copy w-100"},[e._v('\n        Users can upload a CSV or tab-delimited text file with the first column as gene IDs. For RNA-seq data, read count per gene is recommended. Also accepted are normalized expression data based on FPKM, RPKM, or DNA microarray data. iDEP can convert most types of common gene IDs to Ensembl gene IDs, which is used internally for enrichment and pathway analyses. iDEP parses column names to define sample groups. To define 3 biological samples (Control, TreatmentA, TreatmentB) with 2 replicates each, column names should be:\n        Ctrl_1, Ctrl_2, TrtA_1, TrtA_2, TrtB_1, TrtB_2.\n        For factorial design, use underscore "_" to separate factors such as genetic background (wide type vs. mutant:WT vs. Mu) and experimental condition (Ctrl vs. Trt). Currently, only two factors are allowed. To define an 2x2 factorial design, use column names like:\n        WT_Ctrl_1, WT_Ctrl_2, WT_Trt_1, WT_Trt_2, Mu_Ctrl_1, Mu_Ctrl_2, Mu_Trt_1, Mu_Trt_2\n      ')])])]),t("div",{staticClass:"cf ph3 ph5-ns pb5 bg-light-gray black-70 code"},[t("div",{staticClass:"mw9 center"},[t("h1",{staticClass:"fl w-100 mt5 f5 ttu tracked fw6"},[e._v("Step1: Pre Data Process ")]),t("article",{staticClass:"sans-serif pv2 fl w-100 "},[t("h2",{staticClass:"f4 f1-ns fw6 mb2"},[e._v("Pre-process dataset")]),t("div",[e._v("Keep genes with minimal counts per million (CPM) in at least n libraries:")])]),t("div",{staticClass:"cf"},[t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("p",{staticClass:"f5 f4-ns measure lh-copy mt0"},[e._v("\n            Integrated Differential Expression and Pathway analysis (iDEP) of transcriptomic data. See documentation and manuscript. Based on annotation of 69 metazoa and 42 plant genomes in Ensembl BioMart as of 6/4/2017. Additional data from KEGG, Reactome, MSigDB (human), GSKB (mouse) and araPath (arabidopsis). For feedbacks or data contributions (genes and GO mapping of any species), please contact us, or visit our homepage. Send us suggestions or any error message to help improve iDEP.\n          ")])]),t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("section",{staticClass:"f5 f4-ns measure lh-copy mt0"},[t("div",[e._v("\n              Users can upload a CSV or tab-delimited text file with the first column as gene IDs. For RNA-seq data, read count per gene is recommended. Also accepted are normalized expression data based on FPKM, RPKM, or DNA microarray data. iDEP can convert most types of common gene IDs to Ensembl gene IDs, which is used internally for enrichment and pathway analyses. iDEP parses column names to define sample groups. To define 3 biological samples (Control, TreatmentA, TreatmentB) with 2 replicates each, column names should be:\n            ")]),t("div",{staticClass:"code"},[e._v("\n              Ctrl_1, Ctrl_2, TrtA_1, TrtA_2, TrtB_1, TrtB_2.\n            ")]),t("div",[e._v('\n            For factorial design, use underscore "_" to separate factors such as genetic background (wide type vs. mutant:WT vs. Mu) and experimental condition (Ctrl vs. Trt). Currently, only two factors are allowed. To define an 2x2 factorial design, use column names like:\n            ')]),t("div",{staticClass:"code"},[e._v("\n              WT_Ctrl_1, WT_Ctrl_2, WT_Trt_1, WT_Trt_2, Mu_Ctrl_1, Mu_Ctrl_2, Mu_Trt_1, Mu_Trt_2\n            ")])])])])])]),t("div",{staticClass:"cf ph3 ph5-ns pb5 bg-yellow black-70 code"},[t("div",{staticClass:"mw9 center"},[t("h1",{staticClass:"fl w-100 mt5 f5 ttu tracked fw6"},[e._v("Step1: Pre Data Process ")]),t("article",{staticClass:"sans-serif pv2 fl w-100 "},[t("h2",{staticClass:"f4 f1-ns fw6 mb2"},[e._v("Pre-process dataset")]),t("div",[e._v("Keep genes with minimal counts per million (CPM) in at least n libraries:")])]),t("div",{staticClass:"cf"},[t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("p",{staticClass:"f5 f4-ns measure lh-copy mt0"},[e._v("\n            Integrated Differential Expression and Pathway analysis (iDEP) of transcriptomic data. See documentation and manuscript. Based on annotation of 69 metazoa and 42 plant genomes in Ensembl BioMart as of 6/4/2017. Additional data from KEGG, Reactome, MSigDB (human), GSKB (mouse) and araPath (arabidopsis). For feedbacks or data contributions (genes and GO mapping of any species), please contact us, or visit our homepage. Send us suggestions or any error message to help improve iDEP.\n          ")])]),t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("section",{staticClass:"f5 f4-ns measure lh-copy mt0"},[t("div",[e._v("\n              Users can upload a CSV or tab-delimited text file with the first column as gene IDs. For RNA-seq data, read count per gene is recommended. Also accepted are normalized expression data based on FPKM, RPKM, or DNA microarray data. iDEP can convert most types of common gene IDs to Ensembl gene IDs, which is used internally for enrichment and pathway analyses. iDEP parses column names to define sample groups. To define 3 biological samples (Control, TreatmentA, TreatmentB) with 2 replicates each, column names should be:\n            ")]),t("div",{staticClass:"code"},[e._v("\n              Ctrl_1, Ctrl_2, TrtA_1, TrtA_2, TrtB_1, TrtB_2.\n            ")]),t("div",[e._v('\n            For factorial design, use underscore "_" to separate factors such as genetic background (wide type vs. mutant:WT vs. Mu) and experimental condition (Ctrl vs. Trt). Currently, only two factors are allowed. To define an 2x2 factorial design, use column names like:\n            ')]),t("div",{staticClass:"code"},[e._v("\n              WT_Ctrl_1, WT_Ctrl_2, WT_Trt_1, WT_Trt_2, Mu_Ctrl_1, Mu_Ctrl_2, Mu_Trt_1, Mu_Trt_2\n            ")])])])])])]),t("div",{staticClass:"cf ph3 ph5-ns pb5 bg-light-red black-70 code"},[t("div",{staticClass:"mw9 center"},[t("h1",{staticClass:"fl w-100 mt5 f5 ttu tracked fw6"},[e._v("Step1: Pre Data Process ")]),t("article",{staticClass:"sans-serif pv2 fl w-100 "},[t("h2",{staticClass:"f4 f1-ns fw6 mb2"},[e._v("Pre-process dataset")]),t("div",[e._v("Keep genes with minimal counts per million (CPM) in at least n libraries:")])]),t("div",{staticClass:"cf"},[t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("p",{staticClass:"f5 f4-ns measure lh-copy mt0"},[e._v("\n            Integrated Differential Expression and Pathway analysis (iDEP) of transcriptomic data. See documentation and manuscript. Based on annotation of 69 metazoa and 42 plant genomes in Ensembl BioMart as of 6/4/2017. Additional data from KEGG, Reactome, MSigDB (human), GSKB (mouse) and araPath (arabidopsis). For feedbacks or data contributions (genes and GO mapping of any species), please contact us, or visit our homepage. Send us suggestions or any error message to help improve iDEP.\n          ")])]),t("article",{staticClass:"sans-serif pv2 fl w-100 w-50-l pr0 pr2-l"},[t("section",{staticClass:"f5 f4-ns measure lh-copy mt0"},[t("div",[e._v("\n              Users can upload a CSV or tab-delimited text file with the first column as gene IDs. For RNA-seq data, read count per gene is recommended. Also accepted are normalized expression data based on FPKM, RPKM, or DNA microarray data. iDEP can convert most types of common gene IDs to Ensembl gene IDs, which is used internally for enrichment and pathway analyses. iDEP parses column names to define sample groups. To define 3 biological samples (Control, TreatmentA, TreatmentB) with 2 replicates each, column names should be:\n            ")]),t("div",{staticClass:"code"},[e._v("\n              Ctrl_1, Ctrl_2, TrtA_1, TrtA_2, TrtB_1, TrtB_2.\n            ")]),t("div",[e._v('\n            For factorial design, use underscore "_" to separate factors such as genetic background (wide type vs. mutant:WT vs. Mu) and experimental condition (Ctrl vs. Trt). Currently, only two factors are allowed. To define an 2x2 factorial design, use column names like:\n            ')]),t("div",{staticClass:"code"},[e._v("\n              WT_Ctrl_1, WT_Ctrl_2, WT_Trt_1, WT_Trt_2, Mu_Ctrl_1, Mu_Ctrl_2, Mu_Trt_1, Mu_Trt_2\n            ")])])])])])])])}],r={render:s,staticRenderFns:n};a.a=r},a3dU:function(e,a,t){var s=t("ui/k");"string"==typeof s&&(s=[[e.i,s,""]]),s.locals&&(e.exports=s.locals);t("rjj0")("ee68b888",s,!0)},sFlg:function(e,a,t){"use strict";var s=t("NYxO");a.a={computed:Object(s.mapState)(["users"])}},"ui/k":function(e,a,t){a=e.exports=t("FZ+f")(!1),a.push([e.i,'.cf:after,.cf:before{content:" ";display:table}.cf:after{clear:both}.cf{*zoom:1}.fn{float:none}.tracked{letter-spacing:.1em}.lh-title{line-height:1.25}.lh-copy{line-height:1.5}.gray{color:#777}.pv5{padding-top:64px;padding-top:4rem;padding-bottom:64px;padding-bottom:4rem}.ph3{padding-left:16px;padding-left:1rem;padding-right:16px;padding-right:1rem}.mb3{margin-bottom:16px;margin-bottom:1rem}.mt0{margin-top:0}.mt4{margin-top:32px;margin-top:2rem}.ttu{text-transform:uppercase}.f6{font-size:14px;font-size:.875rem}.measure{max-width:30em}@media screen and (min-width:30em){.fl-ns{float:left;display:inline}.w-50-ns{width:50%}.pr4-ns{padding-right:2rem}.ph5-ns{padding-left:4rem;padding-right:4rem}.mt0-ns{margin-top:0}}',""])}});
//# sourceMappingURL=index.0170e2df07f06c9ea49b.js.map